### ISSUE TITLE
DSHawk

### ISSUE THREAD

**Issue by SuuperW** [TASVideos/BizHawk#1688]
I have pushed a branch named DesHawk, as I am integrating DeSmuME into BizHawk. I would appreciate any constructive comments from anyone with knowledge in this area, suggestions about a better way to do something, or potential problems with my approach.

I have also created a DesHawk branch on the DeSmuME repo. (https://github.com/TASVideos/desmume/tree/DesHawk)

**Comment by Spikestuff** [comment]
Well, it's very great at having Graphical Glitches in (only going on a list of Published TASes):
* Pokemon
* Super Monkey Ball: Touch & Roll (Worse offender on this list surprisingly)
* Lego Star Wars

And Dementium has a louder than intended sound in the first minute(ish) of the chapter.

**Comment by zeromus** [comment]
Yeah, you should probably integrate one of the other NDS emulator cores without these bugs. I wish you the best of luck.

**Comment by adelikat** [comment]
Please fork the desmume project, then add the fork as a submodule, so we can get upstream updates easier.  (If you need help with submodules, let me know).  Also preferrably, that fork live in the TASVideos organization, I can set that up

**Comment by zeromus** [comment]
desmume is already on the tasvideos repository. it can be added as a submodule already. upstream is us.

**Comment by adelikat** [comment]
oh right.  rad.

**Comment by alyosha-tas** [comment]
woah desume in bizhawk that sounds great! Hope it works out.

**Comment by Sonia-7** [comment]
I'd also rather see melonDS instead, since I've experienced less emulation bugs there. One example is Etrian Odyssey's text. It displays correctly in melonDS, but in Desmume you need to fix them through a hack (unless if you change the renderer to OpenGL 3.2).

Hopefully it wouldn't be much harder than adding Desmume.

**Comment by NarryG** [comment]
@Sonia-7 Melon does take a bit of work to compile under MSVC (normal toolchain is msys2) but I've done it and it works fine. Melon has a super simple codebase so I would think it'd be easier to port.

**Comment by vadosnaprimer** [comment]
I'd rather we have a thing that's **actually good**. If desmume is added, there will always be demand for melonds anyway, and if than one gets added next, it'd be default, so why not start with a better core altogether? Also melonds is in active developement, while desmume looks [more dead than alive](http://desmume.org/category/releases/), including emulation accuracy improvements.

**Comment by SuuperW** [comment]
Allrighty, I will attempt to use MelonDS. I haven't done anything at all with it yet, but the code looks like at first glance it'll be easier to work with.

**Comment by SuuperW** [comment]
@NarryG you said you got MelonDS to compile under MSVS, can you provide the files or instructions? I have little understanding of how this stuff works.

Or any way I can get to actually working with the code and integrating with BizHawk. I hate how much crap goes into getting C++ to compile. The only way I was able to get DeSmuME to compile was using/copying the pre-existing Windows VS project file.

**Comment by InfamousKnight** [comment]
And I was wrong saying you can only run it through msys2. In the dist folder will be the application.

Wish best of luck to you as well. I dont like the interface of desmume or its bugs. Melonds shouldnt have as many desync issues as well.

**Comment by zeromus** [comment]
It's implausible that melonds will have fewer desync issues if it hasn't been vetted carefully for tasworthiness. Desmume has already been deeply vetted; whatever problems it has have occurred since 0.9.11 and are not likely to be large (but nobody has ever debugged it in detail, to my knowledge). So do not get your hopes up.

**Comment by NarryG** [comment]
Those are copyrighted files. They can't be included for that reason. Desmume uses its own reverse engineered bios which is why it doesn't need one.

**Comment by zeromus** [comment]
desmume's HLE bios could be ported to melonds without too much trouble.

**Comment by SuuperW** [comment]
I have pushed a new branch, DSHawk, which uses MelonDS.

Again, I'd appreciate any constructive feedback.

One specific question I have: Why do the `printf` calls in MelonDS not print to the BizHawk console window? I can get it to by changing the project's type to console application, but I don't recall having to do anything when I was working with DeSmuME.

**Comment by zeromus** [comment]
Desmume has special code in it which attaches stdout to the console. See OpenConsole() in console.cpp

**Comment by vadosnaprimer** [comment]
Hawk related functionality that deviates from default core behavior should be made optional, so the standalone emulator works exactly like before, in terms of both usage and coding. Also avoid magic numbers.

**Comment by MojoJojoDojo** [comment]
problem with melonDS currently is that it runs Mortal Kombat really bad compared to other emulators.
On the good side it emulates the graphics the best out of all other DS emulators including Drastic.
I didnt see any other issues with the emulator though.

**Comment by Spikestuff** [comment]
> problem with melonDS currently is that it runs Mortal Kombat really bad/slow compared to other emulators.

And did you mention that on melonDS?
Cause I'm looking at the issue tracker on melonDS and there's nothing mentioned about it.

Maybe if something was mentioned about it, it would've been taken care of, just like the other issues that as a bonus sometimes actually fixed multiple games at once.
This issue you're not presenting there, instead you're  presenting here doesn't help Arisotura at all.

**Comment by MojoJojoDojo** [comment]
They know about it from the forums  I just wanted to inform people here.

**Comment by zeromus** [comment]
One random game that barely matters and theres no point in tasing isn't really helpful to this discussion. The only reason mortal kombat on the nds matters is because desmume had major problems with it for so long, there will be a lot of small children with too much time on their hands saying melonds will finally execute a fatality on desmume by fixing it. That's why you brought it up here. It isn't relevant.

**Comment by SuuperW** [comment]
> 
> 
> Hawk related functionality that deviates from default core behavior should be made optional, so the standalone emulator works exactly like before, in terms of both usage and coding. Also avoid magic numbers.

@vadosnaprimer Is this in reference to something specific I have done?

**Comment by vadosnaprimer** [comment]
These:
https://github.com/Arisotura/melonDS/pull/527/commits/ce32a666a29146f483a2a1a0459d8f4f2962fc48
https://github.com/Arisotura/melonDS/pull/527/commits/ff0e3eafe5764211bb200c1acce81fe4a940a09e

**Comment by Spikestuff** [comment]
> // Will need to do something about this (and GL_GetProcAddress) for OpenGL support.

Just mentioning this now.
Arisotura has OpenGL as a lesser priority to Soft (or actual DS).

Even though it's pretty functional and can support HiRes I don't recommend implementing it mainly from the encoder's perspective of keeping everything at the native res since every other system except N64 does this.

There'll probably be the HiRes people that wants this feature but honestly it's not worth placing in.

**Comment by SuuperW** [comment]
I'm one of the people that wants HiRes.

**Comment by Spikestuff** [comment]
Fine, I'll mention the other half.
There's a [few feature bugs](https://github.com/Arisotura/melonDS/issues?q=is%3Aissue+is%3Aopen+label%3AOpenGL) that's on that OpenGL that needs to be ironed out if you want to consider it properly.

And apparently people are getting my favourite DeSmuME bug in melonDS which is black scrambled pixels in Pokemon (which has been "removed" in later builds of DeSmuME but only in OpenGL and not soft).

But hey, if you want broken Buzzy Beetles for instance then be my guest.

You shouldn't consider it for a v1 think about it down the line if you want to go that route.

**Comment by DarkLegend1987** [comment]
Hi im a beta tester debbuger  wait for the melón Ds for the bizhawk is a better emu than dsemu me can run many games withdow Bugs or glitch that include Golden sun dark dawn

**Comment by DarkLegend1987** [comment]
Yeah that is the trouble melón Ds use real bios files but like the other cores you develop try to modify some from dsemume to melón Ds because its a better emu to be the hawk client NDS Core

**Comment by DarkLegend1987** [comment]
Yeah the DSemume have bugs with high render games like Golden sun dark dawn, kingdom hearts and other 128mb up to 256mb add as submodule is good because you can upgrade later

**Comment by SuuperW** [comment]
I've just pushed a new .dll file, and I think it's ready to be merged into the master branch. Am I missing anything?

**Comment by vadosnaprimer** [comment]
Please post the archive for efficiency.

**Comment by SuuperW** [comment]
The archive of what?

**Comment by vadosnaprimer** [comment]
Hawk build + core dll.

**Comment by SuuperW** [comment]
[archive.zip](https://github.com/TASVideos/BizHawk/files/4007528/archive.zip)

I fail to see how this makes anything more convenient or efficient, but here ya go.

**Comment by vadosnaprimer** [comment]
It doesn't launch. Run this script https://github.com/TASVideos/BizHawk/blob/DSHawk/Dist/BuildAndPackage_Release.bat

Efficiency is for people who don't have Visual Studio 2019 installed to build bizhawk just to test that single dll.

**Comment by SuuperW** [comment]
That script fails to build. Some error messages I get:

```
BizHawk\Dist>IF NOT EXIST C:\Program Files (x86)\MSBuild\14.0\bin\amd64\nul goto MISSINGMSBUILD
'Files' is not recognized as an internal or external command, operable program or batch file.
```
seems harmless, the file exists

`C:\Program Files (x86)\MSBuild\14.0\bin\amd64\Microsoft.CSharp.CurrentVersion.targets(133,9): warning MSB3884: Could not find rule set file "MinimumRecommendedRules.ruleset". [BizHawk\Version\Version.csproj]`
I have no idea what this rule set is.

`Consoles\Nintendo\NDS\MelonDS_InputPollable.cs(14,29): error CS1043: { or ; expected [BizHawk\BizHawk.Emulation.Cores\BizHawk.Emulation.Cores.csproj]`
There is a { exactly where it says it expects it; similar errors appear in other places in MelonDS code. I have no idea why.

**Comment by vadosnaprimer** [comment]
I'm getting this
![IStOMuE](https://user-images.githubusercontent.com/7092625/71554007-456d3780-2a2a-11ea-9e29-9e0a662f55cb.png)

**Comment by SuuperW** [comment]
I have fixed that bug.

**Comment by InfamousKnight** [comment]
How exciting! Unless linux support can easily be implemented, I might go back to windows just for this.

**Comment by vadosnaprimer** [comment]
![khVWvj0](https://user-images.githubusercontent.com/7092625/71777625-8c6fa580-2fb3-11ea-8113-87be50ba405d.png)

**Comment by vadosnaprimer** [comment]
It's been working for you perfectly fine all this time, right? I can't even get to launching a game yet...

![JM13LoD](https://user-images.githubusercontent.com/7092625/72077969-4004cc80-3309-11ea-88e5-ea3df2bb0f3d.png)

**Comment by SuuperW** [comment]
Yes, it works for me. Where is this latest error you're getting? Can you post the call stack?
Does it show up when you just try to load an NDS ROM?

**Comment by vadosnaprimer** [comment]
The stack is ridiculously useless, just like the error itself is ridiculous. If I uncomment the problematic line is loads fine.

```
 	mscorlib.dll!System.IO.__Error.WinIOError(int errorCode, string maybeFullPath)	Unknown
 	mscorlib.dll!System.IO.__ConsoleStream.Write(byte[] buffer, int offset, int count)	Unknown
 	mscorlib.dll!System.IO.StreamWriter.Flush(bool flushStream, bool flushEncoder)	Unknown
 	mscorlib.dll!System.IO.StreamWriter.Write(char[] buffer, int index, int count)	Unknown
 	mscorlib.dll!System.IO.TextWriter.SyncTextWriter.WriteLine(string value)	Unknown
 	mscorlib.dll!System.Console.WriteLine(string value)	Unknown
>	EmuHawk.exe!BizHawk.Client.EmuHawk.MainForm.FrameBufferResized() Line 1001	C#
 	EmuHawk.exe!BizHawk.Client.EmuHawk.MainForm.Render() Line 2078	C#
 	EmuHawk.exe!BizHawk.Client.EmuHawk.MainForm.ProgramRunLoop() Line 501	C#
 	EmuHawk.exe!BizHawk.Client.EmuHawk.Program.SubMain(string[] args) Line 230	C#
 	EmuHawk.exe!BizHawk.Client.EmuHawk.Program.Main(string[] args) Line 83	C#
 	[Native to Managed Transition]	
 	mscoreei.dll!000007fef38c7b2d()	Unknown
 	mscoree.dll!000007fef39b5b21()	Unknown
 	kernel32.dll!0000000076db59ed()	Unknown
 	ntdll.dll!0000000076fec541()	Unknown
```

Someone else should check this.

**Comment by vadosnaprimer** [comment]
Somehow it also requires `bizhawk\output\melon` dir to exist even if it's **empty**. The game loads fine with that empty dir there (if I comment out the line above) but fails to find the bios if I remove the dir.

**Comment by SuuperW** [comment]
Derp, I should have realized that myself.
I pushed a fix.

**Comment by vadosnaprimer** [comment]
Is it completely impossible to have the firmware files in the dedicated firmware folder? Also I don't understand where it's getting the files from if the dir is empty...

**Comment by SuuperW** [comment]
MelonDS writes to the firmware file. It would be possible to give MelonDS an option to disable this, which would also require making a copy of the writable section in order to load that on reboot. Beyond that it would be a matter of adding methods to set those paths, and replacing all the hard-coded paths. So no, it is not impossible, just easier to do it this way and I didn't think it was important.

Currently the firmware files are being copied to output/melon.

**Comment by InfamousKnight** [comment]
Is there anything else that needs to be done before dshawk branch can be merged to master? Ive heard from yoshirulz that a quick look suggests linux support can easily be added. Been looking forward to that! Unfortunately, we dont know about desync issues yet.

**Comment by SuuperW** [comment]
There are a few things I've found that could be done, but I do not know if the BizHawk admins would consider them required before merging.
1) Savestates do not include the framebuffer, and the next 2 frames after that can in some cases be displayed incorrectly. (Pressing frame advance 3 times may be required before the display is correct.) The first of these 2 frames is because the 3D buffer isn't saved. (3D renderings are displayed the frame after being rendered.) The second, I'm not sure; DeSmuME has some issues here as well, if that's relevant.
2) Opening the Cheats menu crashes. Probably because there the core does not implement a IMemoryDomains.SystemBus. (Although the crashing is an issue with BizHawk, as IMemoryDomains.HasSystemBus does return false.)
3) On that topic, other memory regions besides main memory should be added at some point.
4) It may be desirable to add more firmware settings to the NDS settings dialog.
5) The power button is not emulated.
6) While looking at (1), I noticed that my savestate version-checking code doesn't match pre-existing code. Maybe should be changed. (I only check "minor" version, because old "major" versions are rejected. But some code is checking both.)

About (2), I am not sure how/if a system bus should be implemented.
A) There are two CPUs with different memory mappings. Perhaps we should look at what Action Replays use.
B) The lowest and highest mapped addresses are very far apart, leaving huge empty spaces between. Idk what issues leaving these in would cause for the client. For the core, we'd have to disable some logging when attempting to read/write those addresses.

I am busy with other things, so I may not be doing anything for a while unless it is deemed necessary or I am specifically asked.

**Comment by vadosnaprimer** [comment]
Framebuffer is obtained separately anyway, it's even an option in hawk to have states with or without it.

**Comment by InfamousKnight** [comment]
About linux support, would it be as easy as just compiling it as .so instead of dll given that melonds is cross platform?

**Comment by YoshiRulz** [comment]
@SuuperW I've dealt with point 7 on your list and added point 8.

@InfamousKnight It's still a maybe. melonDS might have chosen to use a different API for their Linux builds, one reason might be naming conventions.

**Comment by vadosnaprimer** [comment]
How about you stop repeatedly bringing up Linux in every bizhawk issue and only focus on it in the issue that is dedicated to it?
https://github.com/TASVideos/BizHawk/issues/1430

**Comment by zeromus** [comment]
Desmume fundamentally supports the savestates without the next few frames rendering being blank or scrambled, but it repeatedly gets wrecked (what happens if the HD rendering settings change between savestate and loadstate?) It's very sensitive and hard to get right. So, it's possible, but it needs to be done deep inside melonDS. @vadosnaprimer  the problem is more the next 2-3 frames than the current frame. You will see how it is very frustrating to tas unless that's perfect.

SystemBus should be arm9, without a doubt. That's basically the "user" cpu, where as the arm7 is meant for the system. However, action replay does manipulate the arm7. It doesnt matter though, since the main memory is mapped to the same range on both buses.

One idea that might help get a lot of mileage is to make "which cpu is systembus" a sync setting and allow the user to change between arm7 and arm9.

"The lowest and highest mapped addresses are very far apart, leaving huge empty spaces between."
We have this situation on ANY 32bit system. It's annoying, but not a problem.
Tools finetuned for the nintendo DS would probably throw off the top nybble of the address to give less space to scroll through since there are essentially 16 repeated memory maps and the top nybble is universally ignored, except to patch the arm9 bios in at the very end of the 32bit address space. But I don't think it's worth worrying about.

**Comment by SuuperW** [comment]
> 
> 
> It's been working for you perfectly fine all this time, right? I can't even get to launching a game yet...
> 
> ![JM13LoD](https://user-images.githubusercontent.com/7092625/72077969-4004cc80-3309-11ea-88e5-ea3df2bb0f3d.png)
> 
> It says "wrong descriptor". Other cores load fine.

Start BizHawk with the log window already open.
I could have MelonDS not attempt to reopen CONOUT$ if the log window isn't open on init, but the log window is already broken anyway and I think that should be addressed first. I've created a new issue, #1807.

**Comment by SuuperW** [comment]
@vadosnaprimer In regards to the #8 you added to my list, what exactly is the problem? I have not observed any issues in my tests.
Also, the default controls I used were copied from other systems' default controls. So are they also bad?

**Comment by vadosnaprimer** [comment]
I haven't added anything.

**Comment by vadosnaprimer** [comment]
With the problematic line:
```
ARM9 BIOS not found
ARM7 BIOS not found
firmware.bin not found
MAC: 00:09:BF:69:15:F8
FW: WIFI CRC16 = GOOD
FW: AP1 CRC16 = BAD
FW: AP2 CRC16 = BAD
FW: AP3 CRC16 = BAD
FW: USER0 CRC16 = GOOD
FW: USER1 CRC16 = BAD
ROM CRC32: B9DF4B8E
chk_size=3387, key=4A4B4A43, wanted=45544359, offset=00000000
chk_size=1693, key=45545041, wanted=45544359, offset=00000000
chk_size=846, key=45454359, wanted=45544359, offset=00000000
chk_size=423, key=454D4343, wanted=45544359, offset=000034E0
chk_size=211, key=45514759, wanted=45544359, offset=00004F50
chk_size=105, key=45533242, wanted=45544359, offset=00005C80
chk_size=52, key=45535543, wanted=45544359, offset=00006310
chk_size=26, key=45544154, wanted=45544359, offset=00006650
chk_size=13, key=45544659, wanted=45544359, offset=000067F0
chk_size=6, key=45544259, wanted=45544359, offset=000067F0
chk_size=3, key=45544454, wanted=45544359, offset=00006850
chk_size=1, key=45544341, wanted=45544359, offset=00006850
chk_size=0, key=45544341, wanted=45544359, offset=00006860
chk_size=0, key=45544359, wanted=45544359, offset=00006870
ROM entry: 02000000 00000001 00000000
Cart ID: 00001FC2
ARM9: offset=00004000 entry=02000800 RAM=02000000 size=001CC6B8
ARM7: offset=001D0800 entry=02380000 RAM=02380000 size=000269A4
CP15Control = 00052078 (00002078->00052000)
Re-encrypting cart secure area
copied logo
CP15Control = 00002078 (00052078->00002000)
PU: region 0 = 04000033 : enabled, 04000000-08000000
PU: region 1 = 0200002D : enabled, 02000000-02800000
PU: region 2 = 027E0021 : enabled, 027E0000-02800000
PU: region 3 = 08000035 : enabled, 08000000-10000000
PU: region 4 = 027C001B : enabled, 027C0000-027C4000
PU: region 5 = 0100002F : enabled, 01000000-02000000
PU: region 6 = FFFF001D : enabled, FFFF0000-FFFF8000
PU: region 7 = 027FF017 : enabled, 027FF000-02800000
PU: CodeCacheable=00000042
PU: DataCacheable=00000042
PU: DataCacheWrite=00000002
PU: CodeRW=05100011
PU: DataRW=15111011
CP15Control = 0005707D (00002078->00057005)
PU region 0: 04000000-08000000, user=00 priv=07
PU region 1: 02000000-02800000, user=70 priv=77
PU region 2: 027E0000-02800000, user=00 priv=00
PU region 3: 08000000-10000000, user=00 priv=03
PU region 4: 027C0000-027C4000, user=00 priv=03
PU region 5: 01000000-02000000, user=00 priv=07
PU region 6: FFFF0000-FFFF8000, user=50 priv=55
PU region 7: 027FF000-02800000, user=00 priv=03
```
Then it crashes.

With the problematic line commented out:
```
ARM9 BIOS not found
ARM7 BIOS not found
firmware.bin not found
MAC: 00:09:BF:4A:77:02
FW: WIFI CRC16 = GOOD
FW: AP1 CRC16 = BAD
FW: AP2 CRC16 = BAD
FW: AP3 CRC16 = BAD
FW: USER0 CRC16 = GOOD
FW: USER1 CRC16 = BAD
ROM CRC32: B9DF4B8E
chk_size=3387, key=4A4B4A43, wanted=45544359, offset=0000000
chk_size=1693, key=45545041, wanted=45544359, offset=0000000
chk_size=846, key=45454359, wanted=45544359, offset=00000000
chk_size=423, key=454D4343, wanted=45544359, offset=000034E0
chk_size=211, key=45514759, wanted=45544359, offset=00004F50
chk_size=105, key=45533242, wanted=45544359, offset=00005C80
chk_size=52, key=45535543, wanted=45544359, offset=00006310
chk_size=26, key=45544154, wanted=45544359, offset=00006650
chk_size=13, key=45544659, wanted=45544359, offset=000067F0
chk_size=6, key=45544259, wanted=45544359, offset=000067F0
chk_size=3, key=45544454, wanted=45544359, offset=00006850
chk_size=1, key=45544341, wanted=45544359, offset=00006850
chk_size=0, key=45544341, wanted=45544359, offset=00006860
chk_size=0, key=45544359, wanted=45544359, offset=00006870
ROM entry: 02000000 00000001 00000000
Cart ID: 00001FC2
ARM9: offset=00004000 entry=02000800 RAM=02000000 size=001CC
ARM7: offset=001D0800 entry=02380000 RAM=02380000 size=00026
CP15Control = 00052078 (00002078->00052000)
Re-encrypting cart secure area
copied logo
CP15Control = 00002078 (00052078->00002000)
PU: region 0 = 04000033 : enabled, 04000000-08000000
PU: region 1 = 0200002D : enabled, 02000000-02800000
PU: region 2 = 027E0021 : enabled, 027E0000-02800000
PU: region 3 = 08000035 : enabled, 08000000-10000000
PU: region 4 = 027C001B : enabled, 027C0000-027C4000
PU: region 5 = 0100002F : enabled, 01000000-02000000
PU: region 6 = FFFF001D : enabled, FFFF0000-FFFF8000
PU: region 7 = 027FF017 : enabled, 027FF000-02800000
PU: CodeCacheable=00000042
PU: DataCacheable=00000042
PU: DataCacheWrite=00000002
PU: CodeRW=05100011
PU: DataRW=15111011
CP15Control = 0005707D (00002078->00057005)
PU region 0: 04000000-08000000, user=00 priv=07
PU region 1: 02000000-02800000, user=70 priv=77
PU region 2: 027E0000-02800000, user=00 priv=00
PU region 3: 08000000-10000000, user=00 priv=03
PU region 4: 027C0000-027C4000, user=00 priv=03
PU region 5: 01000000-02000000, user=00 priv=07
PU region 6: FFFF0000-FFFF8000, user=50 priv=55
PU region 7: 027FF000-02800000, user=00 priv=03
PU: region 1 = 0200002B : enabled, 02000000-02400000
PU region 0: 04000000-08000000, user=00 priv=07
PU region 1: 02000000-02400000, user=70 priv=77
PU region 2: 027E0000-02800000, user=00 priv=00
PU region 3: 08000000-10000000, user=00 priv=03
PU region 4: 027C0000-027C4000, user=00 priv=03
PU region 5: 01000000-02000000, user=00 priv=07
PU region 6: FFFF0000-FFFF8000, user=50 priv=55
PU region 7: 027FF000-02800000, user=00 priv=03
PU: region 2 = 023E0021 : enabled, 023E0000-02400000
PU region 0: 04000000-08000000, user=00 priv=07
PU region 1: 02000000-02400000, user=70 priv=77
PU region 2: 023E0000-02400000, user=00 priv=00
PU region 3: 08000000-10000000, user=00 priv=03
PU region 4: 027C0000-027C4000, user=00 priv=03
PU region 5: 01000000-02000000, user=00 priv=07
PU region 6: FFFF0000-FFFF8000, user=50 priv=55
PU region 7: 027FF000-02800000, user=00 priv=03
PU: DataRW=15115011
PU region 0: 04000000-08000000, user=00 priv=07
PU region 1: 02000000-02400000, user=70 priv=77
PU region 2: 023E0000-02400000, user=00 priv=00
PU region 3: 08000000-10000000, user=00 priv=01
PU region 4: 027C0000-027C4000, user=00 priv=03
PU region 5: 01000000-02000000, user=00 priv=07
PU region 6: FFFF0000-FFFF8000, user=50 priv=55
PU region 7: 027FF000-02800000, user=00 priv=03
unknown ARM9 IO write32 04001060 00000000 01FFF444
unknown ARM9 IO write32 04001064 00000000 01FFF444
unknown ARM9 IO write32 04001068 00000000 01FFF444
unknown save SPI command 01 F0 1
```
Then the game launches fine.

**Comment by YoshiRulz** [comment]
@SuuperW That was me: the key names have been changed in `master` after your branch ("Left" instead of "LeftArrow"), but if you updated now it wouldn't work.

**Comment by SuuperW** [comment]
@vadosnaprimer If you are getting this error even when the log window open from BizHawk start-up (NOT just when you open the ROM; you must have the log window open automatically), then I do not know why you're getting this error.
I will push an update fixing the error which I am able to reproduce, though. Maybe it will fix yours too?

> 
> 
> Framebuffer is obtained separately anyway, it's even an option in hawk to have states with or without it.

I'm using default settings, and have the option enabled (Config -> Rewind & States... -> Save a screenshot in savestates), but I still do not see the savestated frame after loading a savestate. Is this a bug in BizHawk, or is there more to it?

@YoshiRulz Thanks. I'll get the new names from the master branch

**Comment by vadosnaprimer** [comment]
Is it present in a savestate if you open it as a zip?

**Comment by SuuperW** [comment]
Yes.

**Comment by vadosnaprimer** [comment]
@adelikat @zeromus ^

**Comment by zeromus** [comment]
Does it work in other cores on your system?
if YES, then debug your core
if NO, then debug your settings

**Comment by SuuperW** [comment]
I have fixed (hopefully) the errors related to the console, default controls, savestate version checks, implemented the system bus, and fixed most of the savestate display bugs. (The 2nd frame after loading a savestate still has some issues, but I fixed the worst bit.)

**Comment by vadosnaprimer** [comment]
Seems to be loading fine now!

**Comment by Spikestuff** [comment]
Just for the sake of pointing out. I tested a few random games that I had laying about from previous publications/reencodes and three of these titles failed to actually reach menu in the Hawk build, but not upstream melonDS.
The titles (and regions) are as follows:
- Meteos (USA)
- Mario & Luigi - Partners in Time (Euro)
- The Urbz: Sims in the City (USA)

To be a bit more accurate it what I see it's just a white screen.
And depending on the game it'll tank to sub 20fps or while uncapped 400fps+
![image](https://user-images.githubusercontent.com/12412764/74087402-f7932700-4adf-11ea-91bb-b7bd1e204d1d.png)

**Comment by SuuperW** [comment]
@vadosnaprimer I get (what I assume are) the same errors when I try to run that script, as described in my post on this issue on 29 Dec. I have absolutely no idea why they show up or how to debug this problem.

@Spikestuff You're probably not using the bios and/or firmware files in BizHawk. I tested one of the games you mentioned (Meteos) and it reaches the menu for me when the bios files are present.

**Comment by InfamousKnight** [comment]
So how do you compile this? I wanted to test out some stuff too.

**Comment by Spikestuff** [comment]
@SuuperW Yup that's on me. For some reason I was under the impression that the HLE bios was implemented.

**Comment by InfamousKnight** [comment]
How did you manage to get it to build? Im getting the same errors as feos is.

**Comment by vadosnaprimer** [comment]
Build from inside VS.

**Comment by InfamousKnight** [comment]
I'm getting these errors:
Severity	Code	Description	Project	File	Line	Suppression State
Error		The reference assemblies for .NETFramework,Version=v4.6.1 were not found. To resolve this, install the Developer Pack (SDK/Targeting Pack) for this framework version or retarget your application. You can download .NET Framework Developer Packs at https://aka.ms/msbuild/developerpacks	BizHawk.Emulation.DiscSystem			

If  I try to install the runtime library its mentioning, it wont install because its outdated. I dont know how to retarget either.

**Comment by InfamousKnight** [comment]
I got to compile! Tested burnout legends, did a short race, got one take down, and attempted to avoid a crash with different angles, it synced on playback! I also tested slow down and frame and they work well.

**Comment by vadosnaprimer** [comment]
Implemented in https://github.com/TASVideos/BizHawk/commit/c59d8130bbb0cc9d4a05aecdfd428d3cbdc1dd92

**Comment by YoshiRulz** [comment]
For those who care, upstream PR is Arisotura/melonDS#527.

**Comment by vadosnaprimer** [comment]
> I have fixed (hopefully) the errors related to the console

@SuuperW do you happen to remember how you fixed it? I'm having the same issue with my mame core that I updated, and a similar error appeared with `Console.WriteLine`.

**Comment by SuuperW** [comment]
No, and what I did only fixed it in the particular situation I was using. I remember that whether or not it works depends on the order of things happening, not all of which can be reasonably changed.
I ended up deciding that it just wasn't worth the hassle of dealing with Windows' **** system to try to make it work reliably. Also, if you are using `Console.WriteLine`, that works differently from `printf`, which is the method used by MelonDS.

If you can find a way of getting the info you need to show up by changing stuff that shouldn't be changed, I suggest making a new local branch for your own use and don't worry about making it work for everyone. If it's a desired feature for release, I would suggest you re-think that feature unless you're willing to go through a lot of headache to figure this out.
